lib/WasmReader/WasmByteCodeGenerator.cpp

```diff
@@ -1182,62 +1123,59 @@ WasmBytecodeGenerator::EmitReturnExpr(EmitInfo* explicitRetInfo)
+        m_writer->AsmReg1(Js::OpCodeAsmJs::LdUndef, 0);
     }
     else
     {
         EmitInfo retExprInfo = explicitRetInfo ? *explicitRetInfo : PopEvalStack();
-
-        if (m_funcInfo->GetResultType() != retExprInfo.type)
+        if (retExprInfo.type != WasmTypes::Any)
         {
-            throw WasmCompilationException(_u("Result type must match return type"));
-        }
+            if (m_funcInfo->GetResultType() != retExprInfo.type)
+            {
+                throw WasmCompilationException(_u("Result type must match return type"));
+            }
 
-        Js::OpCodeAsmJs retOp = GetReturnOp(retExprInfo.type);
-        m_writer.Conv(retOp, 0, retExprInfo.location);
-        ReleaseLocation(&retExprInfo);
+            Js::OpCodeAsmJs retOp = GetReturnOp(retExprInfo.type);
+            m_writer->Conv(retOp, 0, retExprInfo.location);
+            ReleaseLocation(&retExprInfo);
+        }
     }
-    m_writer.AsmBr(m_funcInfo->GetExitLabel());
+    m_writer->AsmBr(m_funcInfo->GetExitLabel());
 
     SetUnreachableState(true);
 }
 
 EmitInfo
 WasmBytecodeGenerator::EmitSelect()
 {
-    EmitInfo conditionInfo = PopEvalStack();
-    if (conditionInfo.type != WasmTypes::I32)
-    {
-        throw WasmCompilationException(_u("select condition must have I32 type"));
-    }
+    EmitInfo conditionInfo = PopEvalStack(WasmTypes::I32, _u("select condition must have i32 type"));
 
-    Js::ByteCodeLabel falseLabel = m_writer.DefineLabel();
-    Js::ByteCodeLabel doneLabel = m_writer.DefineLabel();
+    Js::ByteCodeLabel falseLabel = m_writer->DefineLabel();
+    Js::ByteCodeLabel doneLabel = m_writer->DefineLabel();
 
-    m_writer.AsmBrReg1(Js::OpCodeAsmJs::BrFalse_Int, falseLabel, conditionInfo.location);
+    m_writer->AsmBrReg1(Js::OpCodeAsmJs::BrFalse_Int, falseLabel, conditionInfo.location);
     ReleaseLocation(&conditionInfo);
 
     EmitInfo falseInfo = PopEvalStack();
-    EmitInfo trueInfo = PopEvalStack();
+    EmitInfo trueInfo = PopEvalStack(falseInfo.type, _u("select operands must both have same type"));
 
     // Refresh the lifetime of the true location
-    m_writer.AsmReg2(GetLoadOp(trueInfo.type), trueInfo.location, trueInfo.location);
+    m_writer->AsmReg2(GetLoadOp(trueInfo.type), trueInfo.location, trueInfo.location);
 
-    m_writer.AsmBr(doneLabel);
-    m_writer.MarkAsmJsLabel(falseLabel);
-
-    if (trueInfo.type != falseInfo.type)
-    {
-        throw WasmCompilationException(_u("select operands must both have same type"));
-    }
+    m_writer->AsmBr(doneLabel);
+    m_writer->MarkAsmJsLabel(falseLabel);
 
     Js::OpCodeAsmJs op = GetLoadOp(trueInfo.type);
 
-    m_writer.AsmReg2(op, trueInfo.location, falseInfo.location);
+    m_writer->AsmReg2(op, trueInfo.location, falseInfo.location);
     ReleaseLocation(&falseInfo);
 
-    m_writer.MarkAsmJsLabel(doneLabel);
-
+    m_writer->MarkAsmJsLabel(doneLabel);
+    
+    if (trueInfo.type == WasmTypes::Any && falseInfo.type != WasmTypes::Any)
```

Comment:
I tried removing that code and then we failed a test.
I remember why I had to add this.
in the case
```js
var res = cond ? unreachable : (i64); // res must be type i64, not Any
var res2 = (i32) + res; // invalid since res is i64 type.
```

I decided to rewrite EmitSelect and I hope the new version is clearer